enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/log/cache/CircularBuffer.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.causalclustering.core.consensus.log.cache;
+
+import static java.lang.Math.floorMod;
+import static java.util.Arrays.fill;
+
+/**
+ * <pre>
+ * Design
+ *
+ * S: start index
+ * E: end index
+ *
+ * When S == E the buffer is empty.
+ *
+ * Examples:
+ *
+ *              S
+ *              v
+ * Empty   [ | | | | | | ]
+ *              ^
+ *              E
+ *
+ *
+ *                S
+ *                v
+ * Size 2  [ | | | | | | ]
+ *                    ^
+ *                    E
+ *
+ *
+ *                 S
+ *                 v
+ * Full   [ | | | | | | ]
+ *               ^
+ *               E
+ *
+ * New items are put at the current E, and then E is moved one step forward (circularly).
+ * The item at E is never a valid item.
+ *
+ * If moving E one step forward moves it onto S
+ * - then it knocks an element out
+ * - and S is also moved one step forward
+ *
+ * The S element has index 0.
+ * Removing an element moves S forward (circularly).
+ *
+ * @param <V> type of elements.
+ */
+public class CircularBuffer<V>
+{
+    private final int arraySize; // externally visible capacity is arraySize - 1
+    private Object[] elementArr;
+
+    private int S;
+    private int E;
+
+    CircularBuffer( int capacity )
+    {
+        if ( capacity <= 0 )
+        {
+            throw new IllegalArgumentException( "Capacity must be > 0." );
+        }
+
+        this.arraySize = capacity + 1; // 1 item as sentinel (can't hold entries)
+        this.elementArr = new Object[arraySize];
+    }
+
+    /**
+     * Clears the underlying buffer and fills the provided eviction array with all evicted elements.
+     * The provided array must have the same capacity as the circular buffer.
+     *
+     * @param evictions Caller-provided array for evictions.
+     */
+    public void clear( V[] evictions )
```

Comment:
It's purely for performance.

Reply:
In that case you could copy into `evictions` with a few calls to `System.arrayCopy` instead of in a for-loop, but it may not be worth writing the extra logic.